### DISCUSSION TITLE
ImageMagick command line crop & remove background

### DISCUSSION THREAD

**Discussion by karantzak** [ImageMagick/ImageMagick#7084] [Help]
### ImageMagick version

7.0.10-19

### Operating system, version and so on

Windows 10

### Description

I would like to automate the process of crop and remove background of an image.

Having an image with black background (see attachment image_1), I would like to perform the following steps:
1. Crop the image and remove empty space around the image (see attachment: image_2) 
2. Remove the black backgroung from the croped image (see attachment: image_3)

![image_1](https://github.com/ImageMagick/ImageMagick/assets/81800494/60738fde-1d2d-4196-8ce2-3d43aea1472c)
![image_2_crop](https://github.com/ImageMagick/ImageMagick/assets/81800494/deac2807-94a2-4f7d-a08c-7016ed22309d)
![image_3_removebg](https://github.com/ImageMagick/ImageMagick/assets/81800494/3c224e45-ff02-4a2d-8c7c-e8618546e704)

Is it possible to automate this process using the magick command-line tool?

Please help

**Comment by snibgo** [comment]
For some reason, I can't download your image. If you zip it and upload the zip file, I will be able to download it.

Your background has no detail, compared to the object. `-edge 1` or similar will find the detail, and `-format %@` will find the bounding box of the detail. You might need to expand the detail slightly with `-morphology dilate`. This gives the crop parameters for your step (1).

If the dark text doesn't merge into the dark background, this will also give a boundary for the object. A flood-fill from top left can create a mask for making the background transparent.

**Comment by karantzak** [comment]
[magick_forum.zip](https://github.com/ImageMagick/ImageMagick/files/14151759/magick_forum.zip)

I am uploading the .zip with the three images. 
Excume me, but I cannot understand the rest of your answer regarding the various parameters,
Could you, please, help me providing the commands that will make my job.
Thanks

**Comment by fmw42** [comment]
Do all your image have the same background? If not, please provide a few more examples

**Comment by karantzak** [comment]
I may have images from different sources. 

- The image, sent to you, is taken by us using a Photo box with a black background (see folder "Black" for more examples).
- There are images in the web that usally have a white background (see examples in folder "White")
- But there are also photos with other background color (see example in folder "grey")

I upload a .zip with the above examples

[images_more.zip](https://github.com/ImageMagick/ImageMagick/files/14156180/images_more.zip)

**Reply by fmw42** [reply]
There are images that have too similar colors for the background in relation to parts or all of the border of your objects (see some of the white background images and the first black background image).  This means that it will be very hard to separate the object from the background by edges or thresholding or any other simple image processing method.  You likely will need to use an AI solution such as at remove.bg to get more consistently good separation (in my opinion).  

Alternately, you could photograph your objects against a **uniform** color background that complements the object color. But that means different colors for each object.

**Comment by karantzak** [comment]
Thank you for your time.
